explorer/explorer_android.go

```diff
@@ -125,24 +171,56 @@ func (f *File) Read(b []byte) (n int, err error) {
 		return 0, err
 	}
 	if n == -1 {
-		n, err = 0, io.EOF
+		return 0, io.EOF
 	}
 	return n, err
 }
 
-func (f *File) Write(b []byte) (n int, err error) {
-	if f == nil || !f.isWritable || f.isClosed {
+func (f *FileReader) Close() error {
+	if f == nil || f.isClosed {
+		return io.ErrClosedPipe
```

Comment:
This comment is more about error handling in general than it is about this particular error case. Can we provide more descriptive errors than `io.ErrorClosedPipe`? I agree that the semantics of that pipe error make sense, but it would be nice to find a way to propagate the underlying java error string into go. What do you think @Inkeliz? Please note that this is a question, not (currently) a request for you to change it.

Reply:
I think the issue is getting the error from Java/Objective-C. I think the best way is creating a custom classe/interface for `File`. That classe/object/interface will have a `error` attribute. So, if  `Read`or `Write` returns `-1` (or `0` then check error.

I think on Objective-C will easier to understand than Java, but the idea is the same:

```
@interface explorer_file:NSObject
@property NSFileHandler* file;
@property NSError* err;
@end 

uint64_t fileRead(CFTypeRef handler, uint8_t *b, uint64_t len) {
        explorer_file * e = (__bridge explorer_file *)handler;
	if (@available(iOS 13, *)) {
		NSData *data = [e.file readDataUpToLength:len error:&e.err]; // << set the error to the interface
		if (data == nil || e.err != nil) {
		return 0;
		}
		[data getBytes:b length:data.length];
		return data.length;
	}
	e.err = NSError("iOS version not supported"); // << set custom error;
	return 0;
}

(char*) checkError(CFTypeRef handler) {
        explorer_file * e = (__bridge explorer_file *)handler;
        return e.err; // get the error from the last call (if it returns 0); it will be null if success;
}
```

I think some similar to Java.

Reply:
This approach sounds good to me.

Reply:
The last commit I add some `getError` to iOS, but I'm not sure if it's working, since I can't trigger the error. Any idea? I think the only way to cause error is by trying to write a really big file (say 1TB) and it will run out of space.

Reply:
I would try simply testing the error return path by artificially generating one. Hack the native code to throw an error and make sure it bubbles up to the Go code.

Reply:
I implement the error return on Android too. Still missing  one error, if the fragment can't be set. I test the Android error and iOS, it seems to work as intended.